### PR TITLE
Feature/raw cargo opts

### PR DESCRIPTION
**Pull request by bazaah**
Adds a parameter 'test_args_raw' that allows users to pass raw args/options to cargo/libtest.

It also refactors how conditional checks are handled: rather than checking all combinations of bools for each endpoint, it builds up a variable, which is passed to `cargo test`. This makes the logic easier to extend and reason about.

Finally, the various `eq` / `ne` conditionals have their syntax altered:
* Changed 'true' (string) => true (bool)
* Swapped arg positions so the correct type coercion [occurs](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/expressions?view=azure-devops#eq)

### CODE REVIEW

File: azure/test.yml [hunk 1]
```diff
@@ -24,33 +25,38 @@ jobs:
...
-        ${{ insert }}: ${{ parameters.envs }}
-  - ${{ if eq('true', parameters.single_threaded) }}:
-    - script: cargo test --all --features "${{ parameters.features }}" -- --test-threads=1
-      displayName: Run tests (single-threaded)
+  # Cargo options go here
+  - script: echo '##vso[task.setvariable variable=options]$(options) --all'
+  - ${{ if eq(true, parameters.features)}}:
+    - script: echo '##vso[task.setvariable variable=options]$(options) --features "${{ parameters.features }}"'
+  # Add any additional conditional libtest options to this check
+  - ${{ if or(eq(true, parameters.single_threaded), eq(true, parameters.test_ignored))}}:
```
Comment (jonhoo) [inline azure/test.yml]: This won't work, as any non-empty string coerces to `true`. This would mean that if someone passes:
```yaml
parameters:
  test_ignored: 'false'
```
(which some of the Azure DevOps docs recommend even though they shouldn't), then that will be considered `true`. Comparing against the string does not have this issue, as booleans are coerced to the appropriate string (`'true'` and `'false'`).

File: azure/test.yml [hunk 2]
```diff
@@ -24,33 +25,38 @@ jobs:
...
-  - ${{ if ne('true', parameters.single_threaded) }}:
-    - script: cargo test --all --features "${{ parameters.features }}"
-      displayName: Run tests
-      env:
-        ${{ insert }}: ${{ parameters.envs }}
-  - ${{ if eq('true', parameters.single_threaded) }}:
-    - script: cargo test --all --features "${{ parameters.features }}" -- --test-threads=1
-      displayName: Run tests (single-threaded)
+  # Cargo options go here
+  - script: echo '##vso[task.setvariable variable=options]$(options) --all'
```
Comment (jonhoo) [inline azure/test.yml]: I think this needs to be `bash`, not `script`, to be cross-platform?

File: azure/test.yml [hunk 3]
```diff
@@ -24,33 +25,38 @@ jobs:
...
+  - ${{ if eq(true, parameters.features)}}:
+    - script: echo '##vso[task.setvariable variable=options]$(options) --features "${{ parameters.features }}"'
+  # Add any additional conditional libtest options to this check
+  - ${{ if or(eq(true, parameters.single_threaded), eq(true, parameters.test_ignored))}}:
+    - script: echo '##vso[task.setvariable variable=options]$(options) --'
+  # Options passed to libtest go here
+  - ${{ if eq(true, parameters.single_threaded) }}:
+    - script: echo '##vso[task.setvariable variable=options]$(options) --test-threads=1'
+  - ${{ if eq(true, parameters.test_ignored)}}:
+    - script: echo '##vso[task.setvariable variable=options]$(options) --ignored'
```
Comment (jonhoo) [inline azure/test.yml]: This is not quite right. `--ignored` runs _only_ ignored tests. So we actually need to invoke `cargo test` a second time if `test_ignored` is on with the same `options` but _also_ `--ignored`.

File: azure/test.yml [hunk 4]
```diff
@@ -24,33 +25,38 @@ jobs:
...
+  # Run tests with the given options
+  - ${{ if eq(true, parameters.raw_opts) }}:
+    - bash: cargo test ${{ parameters.raw_opts }}
```
Comment (jonhoo) [inline azure/test.yml]: I was surprised when I saw this -- I would have expected `raw_opts` to be a way to pass _additional_ arguments, not to replace any arguments caused by other options. That would mean that this:
```yaml
parameters:
  single_threaded: true
  raw_opts: "--foobar"
```
does not actually run the tests single-threaded.

Comment (jonhoo) [reply azure/test.yml]: As an aside to this, what kind of additional parameters are you expecting users to use here?

File: azure/test.yml [hunk 5]
```diff
@@ -24,33 +25,38 @@ jobs:
           vmImage: macOS-10.14
         Windows:
           vmImage: windows-2019
-  ${{ if ne('true', parameters.cross) }}:
-    variables:
+  variables:
+    ${{ if ne('true', parameters.cross) }}:
       vmImage: ubuntu-16.04
+    options: ''
```
Comment (jonhoo) [inline azure/test.yml]: I like this construction, though I'm a little worried about bash being "helpful" in splitting arguments and parsing quoted arguments incorrectly when we do it this way. Specifically because `echo` will get very confused if passed, say, an argument that contains a `'` inside of the `'` you surround its argument with.

**Comment by jonhoo** [comment]
Thanks, this is a cool change! I've left some inline notes :)

File: azure/test.yml (hunk 1 above)
Comment (bazaah) [reply azure/test.yml]: Fair enough, I was getting strange results earlier when comparing strings against strings, which I why I changed it

File: azure/test.yml (hunk 2 above)
Comment (bazaah) [reply azure/test.yml]: Having looked at the failing tests I would agree, pwrsh seems to include the trailing " ' "

File: azure/test.yml (hunk 3 above)
Comment (bazaah) [reply azure/test.yml]: If that is the case (and looking at `cargo test -- -h` that seems correct) than the documentation for "Ignored Tests" [here](https://crate-ci.github.io/azure-pipelines/configuration.html) is erroneous. There is a libtest option (--include-ignored) which seems to do what the documentation suggests is supposed to happen. What is the intended functionality of `test_ignored` ?

File: azure/test.yml (hunk 4 above)
Comment (bazaah) [reply azure/test.yml]: I personally need to use -Z options, which change occasionally.

It was intended that `raw` overrides any other options passed, so as to avoid any weird edge cases down the line. I modeled it off the lxd/ceph raw commands which also override any other args passed

File: azure/test.yml (hunk 5 above)
Comment (bazaah) [reply azure/test.yml]: I agree, but I'm unsure of how else to create an empty string in pipelines.

File: azure/test.yml (hunk 1 above)
Comment (jonhoo) [reply azure/test.yml]: I _think_  the trick here is to use`eq('true', parameters.x)` and `ne('true', parameters.x)` for parameters you _expect_ people to set to `true` if they set them, and `eq('false', parameters.x)` and `ne('false', parameters.x)` for parameters you expect people to set to false.

Comment (jonhoo) [reply azure/test.yml]: See also https://github.com/crate-ci/azure-pipelines/pull/50

File: azure/test.yml (hunk 3 above)
Comment (jonhoo) [reply azure/test.yml]: Why do you believe those docs are erroneous? Following https://github.com/crate-ci/azure-pipelines/pull/49, when you pass `test_ignored: true` then tests with `#[ignored]` are _also_ run as the docs say (that is, we run both `cargo test` _and_ `cargo test -- --ignored`).

File: azure/test.yml (hunk 4 above)
Comment (jonhoo) [reply azure/test.yml]: Hmm, I worry that that'll be too confusing to users. Maybe if we had a way to emit warnings for particular parameter combinations, but we don't really... I think my inclination here would bet o have `raw_opts` default to `--`, and then inject it at:
```console
$ cargo test --all <other cargo flags> $RAW_OPTS <other libtest flags>
```
That way, users can give both cargo and libtest flags in raw_opts. We'd have to document that they _must_ include the `--` in their raw flags though.

As for the combination of raw options with other options, I would be inclined here to just punt on the issue. That is, include raw opts + the options as dictated by parameters. Usually that will be fine, because the user generally controls those parameters, and can just avoid adding additional options.

File: azure/test.yml (hunk 5 above)
Comment (jonhoo) [reply azure/test.yml]: Ah, sorry, I wasn't referring so much to the empty string, but rather to if a user passes, say:
```yaml
parameters:
  features: "foo,'bar baz'"
```
That will currently make `echo` all sorts of confused since the first `'` in there will close the surrounding `'`. Of course maybe the answer here is just to punt on this entirely, but I _do_ think it's important if we introduce raw_opts into `$(options)`, since that is pretty likely to contain single quotes. It might be sufficient to document that it is not allowed to have single quotes, but even that restriction can be hard to work around. This is why the old code was written the way it was rather than constructing a parameter list at runtime :)

File: azure/test.yml (hunk 3 above)
Comment (bazaah) [reply azure/test.yml]: So `test_ignored` should run both ignored and non-ignored tests?

File: azure/test.yml (hunk 4 above)
Comment (bazaah) [reply azure/test.yml]: Okay I'll change it to align with this

File: azure/test.yml (hunk 3 above)
Comment (jonhoo) [reply azure/test.yml]: Correct!

File: azure/test.yml [hunk 6]
```diff
@@ -37,26 +37,27 @@ jobs:
...
-    - script: echo '##vso[task.setvariable variable=options]$(options) --features "${{ parameters.features }}"'
+    - bash: echo '##vso[task.setvariable variable=options]$(options) --features "${{ parameters.features }}"'
+  # Add any raw args
+  - ${{ if eq(true, parameters.raw_opts) }}:
+    - bash: echo '##vso[task.setvariable variable=options]$(options) ${{ parameters.raw_opts }}'
   # Add any additional conditional libtest options to this check
-  - ${{ if or(eq(true, parameters.single_threaded), eq(true, parameters.test_ignored))}}:
-    - script: echo '##vso[task.setvariable variable=options]$(options) --'
+  # Attempts to ensure that any raw args passed do not interfere
+  - ${{ if and( or(eq(true, parameters.single_threaded), eq(true, parameters.test_ignored)), not(endsWith($(options), '--')) ) }}:
```
Comment (jonhoo) [inline azure/test.yml]: I don't think you need the whole `or()` here -- it's fine to add a trailing `--` :)

File: azure/test.yml [hunk 7]
```diff
@@ -24,33 +25,39 @@ jobs:
...
+  - ${{ if eq(true, parameters.single_threaded) }}:
+    - bash: echo '##vso[task.setvariable variable=options]$(options) --test-threads=1'
+  # Run tests with the given options
+  - bash: cargo test $(options)
+    displayName: Run tests
+    env:
+      ${{ insert }}: ${{ parameters.envs }}
+  - ${{ if eq(true, parameters.test_ignored)}}:
+    - bash: cargo test $(options) --ignored
+    displayName: Run tests (ignored)
```
Comment (jonhoo) [inline azure/test.yml]:
```suggestion
    displayName: Run ignored tests
```

Comment (bazaah) [reply azure/test.yml]: 👍

**Comment by bazaah** [comment]
As a side note, is there a pipelines syntax checker?

**Comment by jonhoo** [comment]
Sadly no, not as far as I'm aware. They have a [JSON schema](https://github.com/microsoft/azure-pipelines-vscode/blob/master/service-schema.json) that can be used to validate, but not sure it really helps much with things like debugging templates :( Might be something that's worthwhile to suggest over in the [Azure DevOps Developer Community](https://developercommunity.visualstudio.com/spaces/21/index.html). They're decently responsive there.

File: azure/test.yml (hunk 1 above)
Comment (bazaah) [reply azure/test.yml]: I've corrected all the checks which expect a bool, but might get a string. Hopefully patching this up a little

File: azure/test.yml (hunk 5 above)
Comment (bazaah) [reply azure/test.yml]: I agree we should document it, but I think even with this rough edge the benefits of:

* Not having to multiply bool checks when adding a new option
* The ability to feed cargo exactly which args are required

outweighs it. YMMV 🤷‍♂

Comment (jonhoo) [reply azure/test.yml]: Yeah, that's a good point. Here's my suggestion, based on the fact that `raw_opts` is most likely to cause trouble. Build up _two_ lists, one for cargo args and one for libtest args. Default `raw_opts` to `--`. Then, for `cargo test`, use:
```
- bash: cargo test $(cargo_opts) ${{ parameters.raw_opts }} $(libtest_opts)
```
and the same for the version with `--ignored`:
```
- bash: cargo test $(cargo_opts) ${{ parameters.raw_opts }} $(libtest_opts) --ignored
```
This way `raw_opts` never goes through `echo ''`, and thus quoting shouldn't be an issue, but we still avoid having to multiply the bool checks and retain the ability to provide specific cargo args.

What do you think?

File: azure/test.yml [hunk 8]
```diff
@@ -24,33 +25,39 @@ jobs:
...
+  # Cargo options go here
+  - bash: echo '##vso[task.setvariable variable=options]$(options) --all'
+  - ${{ if eq(true, parameters.features)}}:
+    - bash: echo '##vso[task.setvariable variable=options]$(options) --features "${{ parameters.features }}"'
+  # Add any raw args
+  - ${{ if eq(true, parameters.raw_opts) }}:
+    - bash: echo '##vso[task.setvariable variable=options]$(options) ${{ parameters.raw_opts }}'
+  # Add any additional conditional libtest options to this check
+  # Attempts to ensure that any raw args passed do not interfere
+  - ${{ if and(or(eq('true', parameters.single_threaded), eq('true', parameters.test_ignored)), not(endsWith(parameters.raw_opts, '--'))) }}:
```
Comment (jonhoo) [inline azure/test.yml]: I don't think you need the whole `or()` part of this -- it's fine to add a trailing `--` :)

File: azure/test.yml (hunk 5 above)
Comment (bazaah) [reply azure/test.yml]: I think that solves the issues, I'll give it a bash now, but it might have to wait til tomorrow. Good idea!

Comment (jonhoo) [reply azure/test.yml]: No worries at all -- there's no rush. We'd rather get it right!
I'm heading on vacation later today, so _may_ take a little longer to get back to you for later changes, just FYI :)

Comment (jonhoo) [reply azure/test.yml]: No worries at all -- there's no rush. We'd rather get it right!
I'm heading on vacation later today, so _may_ take a little longer to get back to you for later changes, just FYI :)

Comment (bazaah) [reply azure/test.yml]: enjoy

File: azure/stages.yml [hunk 9]
```diff
@@ -58,6 +59,7 @@ stages:
          single_threaded: ${{ parameters.single_threaded }}
          nightly_feature: ${{ parameters.nightly_feature }}
          features: ${{ parameters.test_features }}
+         raw_opts: ${{ parameters.test_args_raw }}
```
Comment (jonhoo) [inline azure/stages.yml]: I think I'd like to see slightly different names for these. Specifically, let's use `test_arguments` in `stages.yml` and `arguments` in `test` and `tests`.

File: docs/configuration.md [hunk 10]
```diff
@@ -101,6 +101,25 @@ fashion](https://doc.rust-lang.org/book/ch11-02-running-tests.html#running-tests
...
+
+If the pre-baked `cargo test` parameters are not extensible enough for your
+use case, (for example, if you need to pass unstable `-Z` options) you can
+pass `cargo test` custom arguments that are used *in addition* to any others
+set by parameters. This parameter has a syntax:
+
+`<cargo test args> -- <libtest args>`
+
+However, even if you do not include any custom libtest arguments,
+you still **must** include a trailing `--`.
```
Comment (jonhoo) [inline docs/configuration.md]: I think I'd phrase this as:

> Note that even if you include no libtest arguments, you must still include a trailing `--` so that any libtest arguments passed by other configuration options (such as `single_threaded`) will still work correctly.

Comment (bazaah) [reply docs/configuration.md]: I've changed the wording to reflect your suggestion

**Comment by jonhoo** [comment]
I had one thought last night -- do you think that any cargo arguments one passes through this one would also want passed to `cargo check` or to `clippy` (which also compiles the crate)?

I also just realized that we need to also pass these arguments to `coverage`, otherwise the test suite won't run correctly there.

File: docs/configuration.md [hunk 11]
```diff
@@ -101,6 +101,26 @@ fashion](https://doc.rust-lang.org/book/ch11-02-running-tests.html#running-tests
...
+If the pre-baked `cargo test` parameters are not extensible enough for your
+use case, (for example, if you need to pass unstable `-Z` options) you can
+pass `cargo test` custom arguments that are used *in addition* to any others
+set by parameters. This parameter has a syntax:
+
+`<cargo test args> -- <libtest args>`
+
+Note that even if you include no libtest arguments, you must still include
+a trailing `--` so that any libtest arguments passed by other
+configuration options (such as single_threaded) will still work correctly.
```
Comment (jonhoo) [inline docs/configuration.md]:
```suggestion
configuration options (such as `single_threaded`) will still work correctly.
```

**Comment by bazaah** [comment]
> 
> 
> I had one thought last night -- do you think that any cargo arguments one passes through this one would also want passed to `cargo check` or to `clippy` (which also compiles the crate)?
> 
> I also just realized that we need to also pass these arguments to `coverage`, otherwise the test suite won't run correctly there.

I'm not sure, I haven't looked into the overlap between them. It might be problem but I haven't run into issues yet.

**Comment by jonhoo** [comment]
I think it comes down to whether you require those arguments to be passed to compile, or whether they are just "nice to haves". Similarly, for coverage, _can_ tests be run without those additional arguments, or can they not? If they can't then we definitely need to pass the flags to `tarpaulin` for coverage too. If you can't compile without them, they also need to be passed to `cargo check` and `clippy`.
